src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+                .attr('fill', 'none')
+                .call(Color.stroke, geoLayout[d].gridcolor)
+                .call(Drawing.dashLine, '', geoLayout[d].gridwidth);
+        }
+
+        if(constants.fillLayers.indexOf(d) !== -1) {
+            path.attr('stroke', 'none')
+                .call(Color.fill, geoLayout[adj + 'color']);
+        } else if(constants.lineLayers.indexOf(d) !== -1) {
+            path.attr('fill', 'none')
+                .call(Color.stroke, geoLayout[adj + 'color'])
+                .call(Drawing.dashLine, '', geoLayout[adj + 'width']);
+        }
+    });
+};
+
+proto.updateDims = function(fullLayout, geoLayout) {
+    var b = this.bounds;
+    var hFrameWidth = (geoLayout.framewidth || 0) / 2;
+
+    var l = b[0][0] - hFrameWidth;
+    var t = b[0][1] - hFrameWidth;
+    var w = b[1][0] - l + hFrameWidth;
+    var h = b[1][1] - t + hFrameWidth;
+
+    Drawing.setRect(this.clipRect, l, t, w, h);
+
+    this.bgRect
+        .call(Drawing.setRect, l, t, w, h)
+        .call(Color.fill, geoLayout.bgcolor);
+
+    this.xaxis._offset = l;
+    this.xaxis._length = w;
+
+    this.yaxis._offset = t;
+    this.yaxis._length = h;
+};
+
+proto.updateFx = function(fullLayout, geoLayout) {
+    var _this = this;
+    var framework = _this.framework;
+    var gd = _this.graphDiv;
+    var dragMode = fullLayout.dragmode;
+
+    if(_this.isStatic) return;
+
+    function zoomReset() {
+        var viewInitial = _this.viewInitial;
+        var updateObj = {};
+
+        for(var k in viewInitial) {
+            updateObj[_this.id + '.' + k] = viewInitial[k];
+        }
+
+        Plotly.relayout(gd, updateObj);
+        gd.emit('plotly_doubleclick', null);
+    }
+
+    function invert(lonlat) {
+        return _this.projection.invert([
+            lonlat[0] + _this.xaxis._offset,
+            lonlat[1] + _this.yaxis._offset
+        ]);
+    }
+
+    if(dragMode === 'pan') {
+        _this.bgRect.node().onmousedown = null;
+        framework.call(createGeoZoom(_this, geoLayout));
+        framework.on('dblclick.zoom', zoomReset);
+    }
+
+    framework.on('mousemove', function() {
+        var lonlat = _this.projection.invert(d3.mouse(this));
+
+        if(!lonlat || isNaN(lonlat[0]) || isNaN(lonlat[1])) return;
+
+        _this.xaxis.p2c = function() { return lonlat[0]; };
+        _this.yaxis.p2c = function() { return lonlat[1]; };
+
+        Fx.hover(gd, d3.event, _this.id);
+    });
+
+    framework.on('mouseout', function() {
+        Fx.loneUnhover(fullLayout._toppaper);
+    });
+
+    framework.on('click', function() {
+        Fx.click(gd, d3.event);
+    });
+};
+
+proto.makeFramework = function() {
+    var _this = this;
+    var fullLayout = _this.graphDiv._fullLayout;
+    var clipId = 'clip' + fullLayout._uid + _this.id;
+
+    var defGroup = fullLayout._defs.selectAll('g.clips')
+        .data([0]);
+    defGroup.enter().append('g')
+        .classed('clips', true);
```

Comment:
We've got a few of these scattered about... should we just make a `g.clips` from the beginning when we make `fullLayout._defs`?

Reply:
done in https://github.com/plotly/plotly.js/pull/2030/commits/8f9b75aefcb8ac965405cdf1b8a92fd5b40aa62a